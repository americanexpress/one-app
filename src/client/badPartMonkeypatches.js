/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// made a change

// remove document.(open|write|close)
// if `document.write` is called after the page loads, the entire DOM is erased
// we're loading javascript with async and defer, so prevent any issues
if (typeof document !== 'undefined') {
  document.open = function open() { throw new Error('document.open disabled'); };
  document.write = function write() { throw new Error('document.write disabled'); };
  document.writeln = function writeln() { throw new Error('document.writeln disabled'); };
  document.close = function close() { throw new Error('document.close disabled'); };
}

// disable eval
// eval is one of the bad parts, there's a better way to do the thing you want to
[
  'eval',
  'execScript',
].forEach((name) => {
  if (typeof global[name] !== 'function') {
    return;
  }
  global[name] = function monkeypatchedEval() {
    throw new Error(`${name} is disabled`);
  };
});
// there are imlicit eval forms of setTimeout and setInterval (1st arg as string)
// https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout#Parameters
[
  'setTimeout',
  'setInterval',
].forEach((name) => {
  const orig = global[name];
  global[name] = function monkeypatchedEvalForm(cb, ...args) {
    if (typeof cb === 'string') {
      throw new TypeError(`eval form of ${name} used, this is disabled`);
    }
    return Reflect.apply(orig, this, [cb, ...args]);
  };
});

// TODO cover `new Function('var a = 1;')`, any other implicit eval cases?
