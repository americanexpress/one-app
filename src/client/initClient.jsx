/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { browserHistory, Router, matchPromise } from '@americanexpress/one-app-router';
import { setModuleMap } from 'holocron';

import {
  initializeClientStore, loadPrerenderScripts, moveHelmetScripts, loadServiceWorker,
} from './prerender';
import createRoutes from '../universal/routes';

export default async function initClient() {
  try {
    // eslint-disable-next-line no-underscore-dangle
    setModuleMap(global.__CLIENT_HOLOCRON_MODULE_MAP__);
    moveHelmetScripts();

    const store = initializeClientStore();
    const history = browserHistory;
    const routes = createRoutes(store);

    await loadPrerenderScripts(store.getState());

    const { redirectLocation, renderProps } = await matchPromise({ history, routes });

    if (redirectLocation) {
      // FIXME: redirectLocation has pathname, query object, etc; need to format the URL better
      // TODO: would `browserHistory.push(redirectLocation);` and render below, but app stalls
      window.location.replace(redirectLocation.pathname);
      return;
    }

    const {
      __pwa_metadata__: pwaConfig,
      __render_mode__: renderMode = 'hydrate',
    } = global;

    // we want to kick off service worker installation and store sync
    // as early as possible, while not blocking the app from rendering
    // so we let this async function run at its own pace and call it synchronously
    loadServiceWorker({
      config: pwaConfig,
      dispatch: store.dispatch,
    });

    /* eslint-disable react/jsx-props-no-spreading */
    const App = () => (
      <Provider store={store}>
        <Router {...renderProps} />
      </Provider>
    );
    /* eslint-enable react/jsx-props-no-spreading */

    const render = renderMode === 'render' ? ReactDOM.render : ReactDOM.hydrate;

    render(
      <App />,
      document.getElementById('root')
    );
    [...document.getElementsByClassName('ssr-css')]
      .forEach((style) => style.remove());

    // eslint-disable-next-line no-underscore-dangle
    delete global.__INITIAL_STATE__;
    document.getElementById('initial-state').remove();
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error(error);
    // TODO add renderError
  }
}
