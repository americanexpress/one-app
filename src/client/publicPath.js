/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

if (global.__webpack_public_path__) {
  // https://github.com/webpack/webpack/issues/2776#issuecomment-233208623
  // have to make the assignment inside of the compiled code
  // webpack ignores the global variable

  // this var is defined by webpack
  // eslint-disable-next-line no-undef, camelcase -- webpack API
  __webpack_public_path__ = global.__webpack_public_path__;
}
