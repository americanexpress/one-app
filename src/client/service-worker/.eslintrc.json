{
  "extends": "../../../.eslintrc.json",
  "overrides": [
    {
      "files": "**/*.js",
      "rules": {
        // for clearer error reporting, having anonymous functions may inhibit us from
        // collecting a concise stack trace with the proper names of invoked functions
        "prefer-arrow-callback": "off",
        // generally, self is acceptable to use in a service worker script
        // we want to allow this behavior throughout
        "no-restricted-globals": ["off", "self"],
        // we have `@americanexpress/one-service-worker` as part of devDependencies
        // so we disable this rule to allow it to exist as we use it only during build time
        "import/no-extraneous-dependencies": [
          "off",
          {
            "devDependencies": true
          }
        ]
      }
    }
  ]
}
