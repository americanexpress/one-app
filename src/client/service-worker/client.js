/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  on, register, messageContext, messenger,
} from '@americanexpress/one-service-worker';

import { ERROR_MESSAGE_ID_KEY } from './constants';

export default function serviceWorkerClient({
  scriptUrl, scope, onError,
}) {
  // We listen for any messages that come in from the service worker
  on('message', [
    messageContext(),
    messenger({
      [ERROR_MESSAGE_ID_KEY]: onError,
    }),
  ]);

  // as the first basis, we would register the service worker before performing anything else.
  return register(scriptUrl, { scope });
}
