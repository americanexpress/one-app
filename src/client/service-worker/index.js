/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export function importServiceWorkerClient(settings) {
  return import(/* webpackChunkName: "service-worker-client" */ './client')
    // get the default export and invoke with settings
    .then(({ default: serviceWorkerClient }) => serviceWorkerClient(settings));
}

function clearCache() {
  // get the instanced Cache key in browser CacheStorage
  return caches.keys().then(
    // remove cache instances (deleting cache entries in batch)
    (cacheKeys) => Promise.all(
      cacheKeys
        .filter((cacheKey) => cacheKey.startsWith('__sw'))
        .map((cacheKey) => caches.delete(cacheKey))
    ));
}

export function initializeServiceWorker({
  serviceWorker,
  recoveryMode,
  scope,
  scriptUrl,
  webManifestUrl,
  offlineUrl,
  onError,
}) {
  // If the service worker is unavailable, we would not need
  // to call in the chunk since it is not supported in the given browser.
  if ('serviceWorker' in navigator === false) return Promise.resolve();

  // Before we load in the pwa chunk, we can make a few checks to avoid loading it, if not needed.
  return navigator.serviceWorker.getRegistration()
    .then((registration) => {
      // When the service Worker is not enabled (default)
      if (!serviceWorker) {
        // If by any chance a service worker is present, we clear the cache and remove it.
        if (registration) {
          return registration.unregister().then(clearCache).then(() => registration);
        }
        // When there is no registration, nothing further needed to be done.
        return null;
      }

      if (recoveryMode) {
        // Recovery mode is active if the escape-hatch or recovery scripts are enabled
        if (registration) {
          return registration.update().then(clearCache).then(() => registration);
        }
        // When Escape Hatch is active, updating the worker will unregister the worker
        // and make it redundant, any page navigation afterwards should yield no registration.
        return null;
      }

      // Normal operations will load up the library and integrate with the service worker
      return importServiceWorkerClient({
        scope,
        scriptUrl,
        webManifestUrl,
        offlineUrl,
        onError,
      });
    });
}
