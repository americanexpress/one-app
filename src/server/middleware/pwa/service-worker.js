/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { getClientModuleMapCache } from '../../utils/clientModuleMapCache';

import { getServerPWAConfig } from './config';

function processServiceWorkerScript(script) {
  const holocronModuleMap = `'${JSON.stringify(getClientModuleMapCache().browser)}'`;
  return Buffer.from(script.toString().replace('process.env.HOLOCRON_MODULE_MAP', holocronModuleMap));
}

export default function serviceWorkerMiddleware() {
  return function serviceWorkerMiddlewareHandler(req, res, next) {
    const { serviceWorker, serviceWorkerScope, serviceWorkerScript } = getServerPWAConfig();
    if (serviceWorker === false) return next();
    return res
      .type('js')
      .set('Service-Worker-Allowed', serviceWorkerScope)
      .set('Cache-Control', 'no-store, no-cache')
      .send(processServiceWorkerScript(serviceWorkerScript));
  };
}
