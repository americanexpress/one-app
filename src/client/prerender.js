/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { createHolocronStore } from 'holocron';
import { getLocalePack, addErrorToReport } from '@americanexpress/one-app-ducks';
import { compose } from 'redux';
import { createTimeoutFetch } from '@americanexpress/fetch-enhancers';

import createEnhancer from '../universal/enhancers';
import reducer from '../universal/reducers';
import transit from '../universal/utils/transit';
import { initializeServiceWorker } from './service-worker';

export function initializeClientStore() {
  // Six second timeout on client
  const enhancedFetch = compose(createTimeoutFetch(6e3))(fetch);

  const enhancer = createEnhancer();
  /* eslint-disable no-underscore-dangle */
  const initialState = global.__INITIAL_STATE__ !== undefined
    ? transit.fromJSON(global.__INITIAL_STATE__) : undefined;
  const store = createHolocronStore({
    reducer, initialState, enhancer, extraThunkArguments: { fetchClient: enhancedFetch },
  });

  return store;
}

export function loadPrerenderScripts(initialState) {
  const locale = initialState && initialState.getIn(['intl', 'activeLocale']);
  return locale ? getLocalePack(locale) : Promise.resolve();
}

export function moveHelmetScripts() {
  document.addEventListener('DOMContentLoaded', () => {
    const helmetScripts = [...document.querySelectorAll('script[data-react-helmet]')];
    helmetScripts.forEach((script) => document.body.removeChild(script));
    helmetScripts.forEach((script) => document.head.appendChild(script));
  });
}

export function loadServiceWorker({ dispatch, config }) {
  // To handle any errors that occur during installation, we set this handler
  // for dispatching the error back to the server and tie it to the 'message' event.
  const onError = (error) => dispatch(addErrorToReport(error));
  return initializeServiceWorker({
    onError,
    serviceWorker: config.serviceWorker,
    recoveryMode: config.serviceWorkerRecoveryMode,
    scope: config.serviceWorkerScope,
    scriptUrl: config.serviceWorkerScriptUrl,
    webManifestUrl: config.webManifestUrl,
    offlineUrl: config.offlineUrl,
    // in the event of any failure, the app should not crash for non-critical
    // progressive enhancement and report the error back to the server
  }).catch(onError);
}
