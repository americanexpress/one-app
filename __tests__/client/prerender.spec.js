/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { fromJS } from 'immutable';
import { addErrorToReport } from '@americanexpress/one-app-ducks';

import {
  initializeClientStore, loadPrerenderScripts, moveHelmetScripts, loadServiceWorker,
} from '../../src/client/prerender';
import { initializeServiceWorker } from '../../src/client/service-worker';

jest.mock('@americanexpress/one-app-router', () => ({
  browserHistory: 'browserHistory',
}));
jest.mock('../../src/universal/enhancers', () => jest.fn(() => 'enhancer'));
jest.mock('../../src/universal/utils/transit', () => ({
  fromJSON: jest.fn(() => 'initial state'),
}));
jest.mock('holocron', () => ({
  createHolocronStore: jest.fn(),
}));
jest.mock('../../src/universal/reducers', () => 'reducers');

jest.mock('@americanexpress/one-app-ducks', () => ({
  getLocalePack: jest.fn((locale) => Promise.resolve(locale)),
  addErrorToReport: jest.fn(),
}));

jest.mock('../../src/universal/utils/createTimeoutFetch', () => jest.fn(
  (timeout) => (next) => () => next()
    .then((res) => {
      res.timeout = timeout;
      return res;
    })
));
jest.mock('../../src/client/service-worker', () => ({ initializeServiceWorker: jest.fn(() => Promise.resolve()) }));

describe('initializeClientStore', () => {
  beforeAll(() => {
    global.fetch = () => Promise.resolve({ data: 'data' });
  });

  afterAll(() => {
    global.fetch = undefined;
  });


  afterEach(() => {
    // eslint-disable-next-line no-underscore-dangle
    global.__INITIAL_STATE__ = undefined;
  });

  it('should create the store with initial state if it exists', async () => {
    // eslint-disable-next-line no-underscore-dangle
    global.__INITIAL_STATE__ = fromJS({ some: 'state' });
    initializeClientStore();

    const createEnhancer = require('../../src/universal/enhancers');
    const transitFromJson = require('../../src/universal/utils/transit').fromJSON;
    const { createHolocronStore } = require('holocron');
    const {
      extraThunkArguments: { fetchClient },
    } = createHolocronStore.mock.calls[0][0];

    await expect(fetchClient()).resolves.toEqual({ data: 'data', timeout: 6000 });
    // eslint-disable-next-line no-underscore-dangle
    expect(transitFromJson).toHaveBeenCalledWith(global.__INITIAL_STATE__);
    expect(createHolocronStore).toHaveBeenCalledWith({
      enhancer: 'enhancer', initialState: 'initial state', reducer: 'reducers', extraThunkArguments: { fetchClient },
    });
    expect(createEnhancer).toHaveBeenCalled();
  });

  it('should create the store with no initial state if it is undefined', async () => {
    const createEnhancer = require('../../src/universal/enhancers');
    const transitFromJson = require('../../src/universal/utils/transit').fromJSON;
    // need to call mockClear because jest.clearAllMocks() was removed
    // TODO: switch to clearAllMocks when it is added back to jest https://github.com/facebook/jest/issues/2134
    transitFromJson.mockClear();
    const { createHolocronStore } = require('holocron');
    initializeClientStore();
    const {
      extraThunkArguments: { fetchClient },
    } = createHolocronStore.mock.calls[0][0];
    await expect(fetchClient()).resolves.toEqual({ data: 'data', timeout: 6000 });
    expect(transitFromJson).not.toHaveBeenCalled();
    expect(createHolocronStore).toHaveBeenCalledWith({
      enhancer: 'enhancer', initialState: 'initial state', reducer: 'reducers', extraThunkArguments: { fetchClient },
    });
    expect(createEnhancer).toHaveBeenCalled();
  });
});

describe('loadPrerenderScripts', () => {
  it('should load the correct locale pack', () => {
    const initialState = fromJS({ intl: { activeLocale: 'es-ES' } });

    return loadPrerenderScripts(initialState)
      .then((localePack) => {
        expect(localePack).toBe('es-ES');
        const { getLocalePack } = require('@americanexpress/one-app-ducks');
        expect(getLocalePack).toHaveBeenCalledWith('es-ES');
      });
  });

  it('should still resolve if there is no active locale', () => loadPrerenderScripts()
    .then((localePack) => {
      expect(localePack).toBeUndefined();
      const { getLocalePack } = require('@americanexpress/one-app-ducks');
      // need to call clearMock because jest.clearAllMocks() was removed
      // TODO: switch to clearAllMocks when it is added back to jest https://github.com/facebook/jest/issues/2134
      getLocalePack.mockClear();
      expect(getLocalePack).not.toHaveBeenCalled();
    })
  );
});

describe('moveHelmetScripts', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    [...document.querySelectorAll('script')].forEach((scriptTag) => { scriptTag.parentElement.removeChild(scriptTag); });

    jest.spyOn(document, 'addEventListener').mockImplementation(() => { /* noop */ });
  });

  const createScript = ({ src, helmet }) => {
    const script = document.createElement('script');
    script.src = src;
    if (helmet) script.setAttribute('data-react-helmet', true);
    return script;
  };

  it('should register a listener on the document for DOMContentLoaded', () => {
    moveHelmetScripts();
    expect(document.addEventListener).toHaveBeenCalledTimes(1);
    expect(document.addEventListener).toHaveBeenCalledWith('DOMContentLoaded', expect.any(Function));
  });

  it('should only move scripts that were injected by react-helmet, ignoring others', () => {
    [
      { src: 'hello.js', helmet: true },
      { src: 'world.js', helmet: true },
      { src: 'foo.js', helmet: false },
      { src: 'bar.js', helmet: false },
      { src: 'baz.js', helmet: true },
    ]
      .map(createScript)
      .forEach((script) => document.body.appendChild(script));

    moveHelmetScripts();
    document.addEventListener.mock.calls[0][1]();

    expect(document.querySelectorAll('body script')).toMatchSnapshot('non-helmet scripts');
    expect(document.querySelectorAll('head script')).toMatchSnapshot('helmet scripts');
  });

  it('should not expect NodeList to have forEach due to IE', () => {
    [
      { src: 'hello.js', helmet: true },
      { src: 'world.js', helmet: true },
    ]
      .map(createScript)
      .forEach((script) => document.body.appendChild(script));

    moveHelmetScripts();
    jest.spyOn(NodeList.prototype, 'forEach');
    document.addEventListener.mock.calls[0][1]();
    expect(NodeList.prototype.forEach).not.toHaveBeenCalled();
  });
});

describe('loadServiceWorker', () => {
  const dispatch = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should call initializeServiceWorker and resolve', async () => {
    expect.assertions(2);

    await expect(loadServiceWorker({ dispatch, config: {} })).resolves.toBeUndefined();
    expect(initializeServiceWorker).toHaveBeenCalledTimes(1);
  });

  it('should not crash the application on failure nor should loadServiceWorker reject', async () => {
    expect.assertions(4);

    const error = new Error('ooops');
    initializeServiceWorker.mockImplementationOnce(() => Promise.reject(error));

    await expect(loadServiceWorker({ dispatch, config: {} })).resolves.toBeUndefined();
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(initializeServiceWorker).toHaveBeenCalledTimes(1);
    expect(addErrorToReport).toHaveBeenCalledWith(error);
  });
});
