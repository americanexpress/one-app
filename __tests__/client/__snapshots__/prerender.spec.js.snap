// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moveHelmetScripts should only move scripts that were injected by react-helmet, ignoring others: non-helmet scripts 1`] = `
NodeList [
  <script
    src="foo.js"
  />,
  <script
    src="bar.js"
  />,
]
`;

exports[`moveHelmetScripts should remove helmet scripts from the body and only append helmet scripts to the head if react-helmet has not already injected them: head helmet scripts 1`] = `
Array [
  <script
    data-react-helmet="true"
    src="hello.js"
  />,
]
`;
